src/parcsr_mv/par_csr_matop_device.c

```diff
@@ -1328,6 +1328,14 @@ hypre_ParCSRMatrixDropSmallEntriesDevice( hypre_ParCSRMatrix *A,
    HYPRE_Real      *elmt_tols_diag = NULL;
    HYPRE_Real      *elmt_tols_offd = NULL;
 
+   /* Exit if tolerance is zero */
+   if (tol < HYPRE_REAL_MIN)
```

Comment:
We can prune numerical zeros by setting tol = 0.0. With this change, it becomes a no-op.

Reply:
I would like to hear what others think about it.

Reply:
@rfalgout @oseikuffuor1 In your opinion, What should be the behavior of `hypre_ParCSRMatrixDropSmallEntries` when `tol = 0.0` is passed in?
1. Exit the function
2. Try to remove coefficients that are actually equal to zero from the sparsity pattern

Reply:
It appears the behavior of the parent function [hypre_ParCSRMatrixDropSmallEntries](https://github.com/hypre-space/hypre/blob/58d1abbc274a550da5a1c34c536e2d4a364c1920/src/parcsr_mv/par_csr_matop.c#L6215) returns hypre_error_flag if (tol <=0.0). This means we cannot prune numerical zeros anyway with this function. As Rui Peng noted, I think it is a good idea to be able to do that (I thought we were able to do this), but the change would have to happen at a higher level than this device code (i.e. in par_csr_matop.c)

Reply:
Good point, Daniel! I'm wondering if we should have a different function for the purpose of pruning zeros from the sparsity pattern. That would allow the DropSmallEntries function to behave as no-op in the case of `tol = 0.0` and I think that's important because it would provide an efficient way to turn off the dropping of small entries. I say efficient because in the `tol = 0` scenario, the function would just exit.

Reply:
Another idea is: have a single [hypre_ParCSRMatrixDropSmallEntries](https://github.com/hypre-space/hypre/blob/58d1abbc274a550da5a1c34c536e2d4a364c1920/src/parcsr_mv/par_csr_matop.c#L6215) function, make it drop numerical zeros when `tol = 0` and make it a no-op when `tol < 0`

Reply:
After looking through some of the calling code, I am thinking that your last idea @victorapm is the right way to go, i.e., have a separate routine for dropping zero entries.  This gives the most flexibility since: it allows a no-op; it still allows us to prune zeros from matrices if we want to; and it also lets us keep the zeros (we may want to have a fixed graph structure for some reason, even if some of the entries are zero).

Reply:
Note that by "last idea" in my previous comment, I really meant the idea before your last. :)  So, no `tol<0` option.

Reply:
If I read correctly, I like the idea of tol < 0 being no-op and tol = 0.0 to prune numerical zeros. I was wondering relaxing the if-condition in the parent function will do so. At least the GPU code seems fine which I checked yesterday. Thanks for the discussion.

Reply:
OK. I guess Rob referred to the other idea Victor suggested...

Reply:
I think if we are going with Victor's former suggestion (i.e. new functionality to specifically sparsity by dropping zeros), then we need not change anything in the parent function, and this new code under discussion in the device code should simply be removed. I believe that will make both host and device code behave the same way (i.e. a no-op if tol <= 0.0). Then we can add new functionality for dropping zeros.

Reply:
Agreed. The new function, say named `hypre_ParCSRMatrixDropZeros`, is essentially the same as this one, removing the `if (tol <=0.0) return` of course, at least for the GPU code.

Reply:
Thanks everyone for the inputs! If we have a`hypre_ParCSRMatrixDropZeros`, how many places in hypre would we need to include calls to it? Otherwise, if we go with the single drop function idea with no-op given by `tol < 0`, we would not have to include calls to DropZeros throughout the code. Hmmm, that's tricky! There are advantages in both approaches... Nice discussion!

Reply:
It doesn't look like the current code would remove zeros anyway with `tol=0`.  The thrust calls deep down in `csr_matop_device.c` use `less_than`.  Maybe I'm misreading it.  Anyway, Daniel's suggestion is probably best for now so we can discuss this more.  It might well be that a special routine for dropping zeros can be done by simply calling a version of this routine that uses the `tol<0` approach, or we can skip the special routine altogether.  We would want to minimize writing similar code as much as possible in either case.

Reply:
One other comment about the `tol<0` no-op idea.  This would require the calling code to know ahead of time that it wants a no-op (it requires passing in a negative number and most of these tolerances are assumed to be `>=0`).  If that's the case and you have an extra `if` anyway, why not just skip calling the function altogether instead of passing a negative tolerance?  I think this needs more thought.

Reply:
I think we missed taking absolute values in `csr_matop_device.c`? Oops, this is a bigger problem..

And Rob was right on the `less_than`

Reply:
I was wondering about the absolute value question, but I was guessing you intended to only pass matrices with non-negative entries to it (e.g., strength matrices).  Thanks!

Reply:
It should be a bug. We should add `thrust::make_transform_iterator(A_data,  absolute_value<HYPRE_Real>())`. Sorry.

Reply:
Sounds great! Shall I leave `hypre_ParCSRMatrixDropSmallEntriesDevice` as is in this branch and open another branch to continue working on the dropping strategy, e.g., adding a new `hypre_ParCSRMatrixDropZeros`, make adjustments to `csr_matop_device.c` as talked above and possibly others?

Reply:
Hi all, so I'll make changes related to matrix truncation in another PR. Thanks!